src/topology/config/component.rs

```diff
@@ -0,0 +1,180 @@
+{
+    swap_out: ConfigSwapOut,
+
+    pub component: T,
+}
+
+impl<T> ComponentConfig<T>
+where
+    T: 'static + Sized,
+    inventory::iter<ComponentBuilder<T>>:
+        std::iter::IntoIterator<Item = &'static ComponentBuilder<T>>,
+{
+    /// Returns a sorted Vec of all plugins registered of a type.
+    pub fn types() -> Vec<String> {
+        let mut types = Vec::new();
+        for definition in inventory::iter::<ComponentBuilder<T>> {
+            types.push(definition.name.clone());
+        }
+        types.sort();
+        types
+    }
+
+    /// Returns an example config for a plugin component as a toml::Value.
+    pub fn example_value_for(type_str: &str) -> Result<Value, String> {
+        match inventory::iter::<ComponentBuilder<T>>
+            .into_iter()
+            .find(|t| t.name.as_str() == type_str)
+        {
+            Some(b) => match (b.example_value)() {
+                Ok(d) => Value::try_from(&ConfigSwapOut {
+                    type_str: type_str.to_owned(),
+                    nested: d,
+                })
+                .map_err(|e| format!("{}", e)),
+                Err(e) => Err(e),
+            },
+            None => Err(format!("unrecognized type '{}'", type_str,)),
+        }
+    }
+}
+
+impl<T> Serialize for ComponentConfig<T>
+where
+    T: 'static + Sized,
+    inventory::iter<ComponentBuilder<T>>:
+        std::iter::IntoIterator<Item = &'static ComponentBuilder<T>>,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.swap_out.serialize(serializer)
+    }
+}
+
+impl<'de, T> Deserialize<'de> for ComponentConfig<T>
+where
+    T: 'static + Sized,
+    inventory::iter<ComponentBuilder<T>>:
+        std::iter::IntoIterator<Item = &'static ComponentBuilder<T>>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let swap_out = deserializer.deserialize_map(ConfigSwapOut::new())?;
+        match inventory::iter::<ComponentBuilder<T>>
+            .into_iter()
+            .find(|t| t.name == swap_out.type_str)
+        {
+            Some(b) => match (b.from_value)(swap_out.nested.clone()) {
+                Ok(c) => Ok(Self {
+                    swap_out: swap_out,
+                    component: c,
+                }),
+                Err(e) => Err(Error::custom(format!(
+                    "failed to parse type `{}`: {}",
+                    swap_out.type_str, e,
+                ))),
+            },
+            None => Err(Error::custom(format!(
+                "unrecognized type '{}'",
+                swap_out.type_str,
+            ))),
+        }
+    }
+}
+
+/// Constructs a plugin instance from a toml::Value of its configuration.
+type ComponentFromValue<T> = fn(Value) -> Result<T, String>;
+
+/// Returns a configuration example of a plugin as a toml::Value.
+type ComponentExampleValue = fn() -> Result<Value, String>;
```

Comment:
It'd be nice if we could derive the examples from the config metadata we already have for building documentation instead of writing code to generate one.

Reply:
Yeah I'm going to remove the example config part, I added it as a sort of showcase for stuff we can add but we don't need it right now. If we introduce this stuff later we can make it opt-in for plugins as they register themselves.